hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java

```diff
@@ -139,19 +151,96 @@ public TransactionMetadata preHandleUpdateAccount(TransactionBody txn) {
 
     @Override
     /** {@inheritDoc} */
-    public TransactionMetadata preHandleCryptoTransfer(TransactionBody txn) {
-        throw new NotImplementedException();
+    public TransactionMetadata preHandleCryptoTransfer(@Nonnull final TransactionBody txn) {
+        Objects.requireNonNull(txn);
+        final var op = txn.getCryptoTransfer();
+        final var payer = txn.getTransactionID().getAccountID();
+        final var meta = new SigTransactionMetadata(accountStore, txn, payer);
+
+        for (TokenTransferList transfers : op.getTokenTransfersList()) {
+            final var tokenMeta = tokenStore.getTokenMeta(transfers.getToken());
+
+            for (AccountAmount accountAmount : transfers.getTransfersList()) {
+                if (!tokenMeta.failed()) {
+                    final var isUnapprovedDebit =
+                            accountAmount.getAmount() < 0 && !accountAmount.getIsApproval();
+                    if (isUnapprovedDebit) {
+                        meta.addNonPayerKey(accountAmount.getAccountID());
+                    } else {
+                        meta.addNonPayerKeyIfReceiverSigRequired(
+                                accountAmount.getAccountID(), INVALID_TRANSFER_ACCOUNT_ID);
+                    }
+                }
+            }
```

Comment:
It seems[ these checks](https://github.com/hashgraph/hedera-services/blob/9bf735718c899586a2839c252d11d98316b77567/hedera-node/hedera-mono-service/src/main/java/com/hedera/services/sigs/order/SigRequirements.java#L1294) to decide if the error status can be ignored in the account lookup from `SigRequirements` are missing ?